File: src/pin.rs
use uuid::Uuid;

pub struct Pin {
    pub id: Uuid,
    pub neighbors: Vec<Uuid>,
    pub state: i32,
}

impl Pin {

    pub fn new(id: Uuid) -> Self {
        Pin { id, neighbors: Vec::new(), state: 0 }
    }

    pub fn set_state(&mut self, new_state: i32) {
        self.state = new_state;
    }

    pub fn reset(&mut self) {
        self.state = 0;
    }
    
}

File: src/vpga_graph/pin.clj
(ns vpga-graph.pin)

(defrecord Pin [id state neighbor-pin-ids])

(defn generate
  "Generates an unconnected pin"
  []
  (Pin. "" 0 []))

(defn reset
  "Resets a pin's state to off"
  [pin]
  (assoc pin :state 0))

(defn neighbors
  "Returns the neighbor pin-ids from the given Pin"
  [pin]
  (:neighbor-pin-ids pin))

(defn connect-to
  "Connects the given pin to another given pin"
  [pin other-pin]
  (let [pin-neighbors (:neighbor-pin-ids pin)
        updated-neighbors (conj pin-neighbors (:id other-pin))]
    (assoc pin :neighbor-pin-ids updated-neighbors)))